README.md
# tenmado

origin http://www.interq.or.jp/libra/oohara/tenmado/index.html
license http://metadata.ftp-master.debian.org/changelogs/main/t/tenmado/tenmado_0.10-2_copyright


tenmado-0.10-src/ChangeLog
2012-01-31  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.10
	* LICENSE, const.h: fixed copyright string

2012-01-30  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.9
	* INSTALL, README, README-ja-utf8, configure.ac: removed
	  the libdb3 support
	* configure.ac: use AC_DEFINE after AC_ARG_WITH, not in it
	* configure.ac, Makefile.am, happy-L/Makefile.am,
	  libtenm/Makefile.am: add the return value of sdl-config --libs
	  (which usually contains a -l flag) to LIBS, not AM_LDFLAGS
	* Makefile.am, happy-L/Makefile.am, libbac/Makefile.am,
	  libtenm/Makefile.am: removed -Wconversion because it is too noisy
	* Makefile.am, configure.ac: use the tar-ustar option of automake
          instead of setting AMTAR
	* configure.ac: fixed comments so that configure doesn't say
	  "process this file with autoconf"
	* Makefile.am, happy-L/Makefile.am, libbac/Makefile.am,
	  libtenm/Makefile.am: fixed comments so that Makefile doesn't say
	  "process this file with automake"
	* configure.ac: added AC_CHECK_FUNCS([strerror]) because
	  libbac/bac_load.c wants it
	* libbac/bac_lexical.l: added added %option no* to fix the
	  -Wunused-function warning
	* libbac/bac_lexical.h: added yylex_destroy() declaration
	* libbac/bac_lexical.h, libbac/bac_lexical.l, libbac/bac_parser.y,
	  libbac/bac_string_to_array.c: use %option yylineno instead of
	  counting the line number manually
	* happy-L/close-system.c: removed the variable time_no_escape
	  in the function close_system_ball_new() because it is unused
	* happy-L/ending.c: fixed the return value of ending()
	* happy-L/ending.c: fixed an error message

2009-11-14  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.8
	* libtenm/tenm_timer.c: uses an empty loop for maximum accuracy
	* libbac/bac_truncate.c: fixed the error message when remove()
	  failed

2004-11-25  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.7
	* happy birthday, L!
	* happy-L/normal-enemy.c: added sanity checks after tenm_sqrt()
	* libtenm/tenm_input.c: really fixed the default value of neutral_range
	* tenmado.6: command line options use a minus sign, not a hyphen
	* happy-L/result.c: reset timer at the beginning of the result report
	* README: fixed typos

2003-11-25  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.6
	* happy birthday, L!
	* happy-L/player-shot.c, libtenm/tenm_table.c: made sure that the
	  action of the object that is added by the player (the player shot)
	  is skipped only once (the frame when that object is added)
	* happy-L/midnight.c, happy-L/overrun.c, happy-L/disconnection.c,
	  happy-L/fud.c, happy-L/last-boss.c: don't modify my->attr or
	  my->hit_mask in my->hit(), or the player shot may fly through
	  the enemy
	* Makefile.am, README-ja-sjis: added \r before each \n
	* Makefile.am, happy-L/Makefile.am, libtenm/Makefile.am: added
	  -Wno-long-long to CC_WARNINGS
	* happy-L/player.c: int/double consistency in the boundary check
	* happy-L/player-shot.c: your shot has no longer the same color
	  as an enemy's shot
	* libtenm/tenm_input.c: set default neutral_range to 8192
	* main.c: set neutral_range to 8192
	* libbac/bac_array.h, libbac/bac_entry.h, libtenm/tenm_input.c,
	  libtenm/tenm_object.c, libtenm/tenm_object.h,
	  libtenm/tenm_primitive.h, libtenm/tenm_table.c: fixed pointers
	  to a function
	* happy-L/player-shot.c: fixed typo in an error message
	* libbac/bac_entry.h: includes config.h
	* README, happy-L/tutorial.c: fixed typo
	* happy-L/image.c: removed the unnecessary slash from the path of
	  image files
	* THANKS: updated the e-mail address of shinichiro.h
	* happy-L/explosion.c: int/double consistency in the initial value
	* happy-L/normal-enemy.c: don't decrement the "no escape" timer
	  if it is not positive
	* happy-L/disconnection.c, happy-L/fud.c: a dead enemy has low
	  priority when we draw the window
	* happy-L/explosion.c: explosion has low priority when we draw
	  the window
	* libtenm/tenm_table.c: added sanity check to
	  tenm_table_detect_collision() to handle hit() of an object that
	  deletes the object which collides with it
	* happy-L/default/score-*.txt: updated

2003-01-16  Oohara Yuuma  <oohara@libra.interq.or.jp>

	* version 0.5
	* libdb3 is now optional (still recommended and default)
	* happy-L/disconnection.c, happy-L/fud.c, happy-L/image/*-damaged.png,
	  happy-L/image.c, happy-L/intuition.c, happy-L/last-boss.c,
	  happy-L/midnight.c, happy-L/normal-enemy.c, happy-L/overrun.c,
	  happy-L/solution.c: better enemy damaging graphic
	* configure.ac, happy-L/high-score.c, INSTALL, libbac/bac_load.c,
	  libbac/bac_load.h, libbac/bac_save.c, libbac/bac_truncate.c:
	  added (deprecated) --without-libdb3 support
	* configure.ac, happy-L/high-score.c, happy-L/image.c,
	  happy-L/result.h, INSTALL, libbac/bac_dump.h, libbac/bac_entry.h,
	  save/Makefile.am, util.c: added (dangerous) --without-posix support
	* score.c: reset extend_next whenever a game starts
	* happy-L/intuition.c, happy-L/laser.c, happy-L/normal-shot.c:
	  an enemy shoots at a random direction if it aims at you and if it
	  and you are at the exacely same point
	* happy-L/midnight.c: don't call midnight_bit_signal_to_bit() if
	  the bit is already dead
	* happy-L/last-boss.c: surprise_laser_act() uses laser_new() instead
	  of laser_point_new()
	* libtenm/tenm_timer.c: added clock accuracy auto-detection
	* libtenm/tenm_object.c: the object name is now duplicated,
	  no longer (const char *) but (char *)
	* libbac/bac_array.h: fixed sanity check in bac_array_delete (s/1/i/)
	* README, option.c, tenmado.6: fixed the description of --dump
	  and --undump
	* README: documented extend (additonal ship)
	* README-ja-utf8: new file, Japanese translation
	* README-ja-sjis: new file, generated from README-ja-utf8
	* libbac/README: added more nonsense
	* happy-L/high-score.c, libtenm/tenm_input.c, libtenm/tenm_math.c,
	  libtenm/tenm_table.c, main.c, util.c, util.h: fixed memory leaks
	* libbac/bac_lexical.h, libbac/bac_lexical.l,

tenmado-0.10-src/loop.c
/* $Id: loop.c,v 1.53 2002/12/25 12:30:42 oohara Exp $ */

#include <stdio.h>

#include "tenm_object.h"
#include "pause.h"
#include "tenm_table.h"
#include "player.h"
#include "score.h"
#include "ship.h"
#include "chain.h"
#include "hit-status.h"
#include "stage.h"
#include "tenm_timer.h"
#include "tenm_input.h"
#include "scheduler.h"
#include "const.h"
#include "tenm_graphic.h"
#include "result.h"
#include "esc-ok.h"
#include "option.h"

#include "loop.h"

/*
static int draw_all_mass(tenm_object *my, int n);
*/

/* returns 1 if the program should quit, 0 if not */
int
main_loop(int tutorial)
{
  /* for debugging only */
  int t_start = 0;

  int i;
  int t = 0;
  int frame_passed = 0;
  int quit_needed = 0;
  const option *op = NULL;
  tenm_object *player;

  op = get_option();
  if (op == NULL)
  {
    fprintf(stderr, "main_loop: get_option failed\n");
    /* we should have checked the options */
    return 1;
  }

  clear_pause();

  /* we don't need to pause if it is a tutorial */
  if (tutorial == 0)
    tenm_set_focus_handler((void (*)(int)) pause_by_mouse);
  else
    tenm_set_focus_handler((void (*)(int)) NULL);

  tenm_table_clear_all();
  player = player_new(tutorial);

  clear_score();
  set_ship(5);
  clear_chain();
  clear_hit_status();

  if (tutorial == 0)
  {
    if ((op->stage >= 1) && (op->stage <= 5))
      set_stage(op->stage);
    else
      set_stage(1);
  }
  else
  {  
    set_stage(7);
  }
  
  tenm_timer_reset();

  while (1 == 1)
  {
    /* quit the program if a SDL_QUIT event happened 
     * (for example, if a SIGINT signal (sent by Ctrl+c) is received)
     */
    if (tenm_event_handle() != 0)
    {
      quit_needed = 1;
      break;
    }
    /* back to the title if ESC is pressed  */
    if (tenm_get_key_status() & 32)
    {
      if (get_esc_ok())
      {
        set_esc_ok(0);
        break;
      }
    }
    else
    {
      set_esc_ok(1);
    }

    /* pause if it is a real game and if p is pressed
     * we must check "tutorial" first to suppress unnecessary
     * pause message
     */
    if ((tutorial == 0) && (do_pause(tenm_get_key_status() & 64)))
    {
      frame_passed++;
      /* this wait is necessary to save CPU time */
      tenm_wait_next_frame();
      continue;
    }

    for (i = 1; i <= 30; i++)
    {
      if (tenm_table_detect_collision(player) != 0)
        player = NULL;
      if (tenm_table_move(player, 30) != 0)
        player = NULL;
    }

    /* scheduler() must be called before tenm_table_do_action()
     * to clear action_needed flag correctly */
    if (scheduler(t + t_start) == SCHEDULER_NEXT_STAGE)
    {
      clear_chain();
      /* t is incremented at the end of the loop */
      t = -1;
      /* "stage 6" means ending, "stage 7" means tutorial */
      if (get_stage() > 5)
        break;
      /* you can't go to the next stage if you use --stage option */
      if ((op->stage >= 1) && (op->stage <= 5))
        break;
      add_stage(1);
    }

    if (tenm_table_do_action(player) != 0)
      player = NULL;
    calculate_chain();


    /* pure white is bad for your eyes */
    tenm_clear_window(tenm_map_color(253, 245, 230));

    tenm_table_draw(player);
    show_chain();
    show_hit_status();
    show_score();
    show_ship();

    /* for those who want to see the world as it is */
    /* note that
     * (1) the player is not in the table
     * (2) this is slow
     */
    /*
    tenm_table_apply_all((int (*)(tenm_object *, int)) draw_all_mass, 0);
    */

    tenm_redraw_window();

    tenm_wait_next_frame();
    if (op->slow != 0)
    {
      /* wait one more frame */
      tenm_wait_next_frame();
    }

    if (player == NULL)
      break;
    t++;
    frame_passed++;
  }

  if (frame_passed > 0)
  {
    printf("average fps: %f\n", tenm_calculate_fps(frame_passed));
    fflush(stdout);
  }

  tenm_table_clear_all();
  if (player != NULL)
    tenm_object_delete(player);

  if ((!(tutorial)) && (!(quit_needed)))
  {
    /* note that show_result() does not use table */
    if (show_result() != 0)
      quit_needed = 1;
  }

  return quit_needed;
}

#if 0
static int
draw_all_mass(tenm_object *my, int n)
{
  /* sanity check */
  if (my == NULL)
    return 0;

  if (my->mass == NULL)
    return 0;
  if (my->attr == 0)
    return 0;

  if (tenm_draw_mass(my->mass, tenm_map_color(0, 0, 0)) != 0)
  {
    fprintf(stderr, "draw_all_mass: tenm_draw_mass failed (%d)\n",
            my->table_index);
  }

  return 0;
}
#endif /* 0 */
